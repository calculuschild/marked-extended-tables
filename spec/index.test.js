import { marked } from 'marked';
import extendedTable from '../src/index.js';

function trimLines(s) {
  return s.split('\n').map(l => l.trim()).join('\n');
}

describe('extended-table', () => {
  beforeEach(() => {
    marked.setOptions(marked.getDefaults());
  });

  test('Column Spanning', () => {
    marked.use(extendedTable());
    expect(marked(trimLines(`
      | H1      | H2      | H3      |
      |---------|---------|---------|
      | This cell spans 3 columns |||
    `))).toMatchSnapshot();
  });

  test('Row Spanning', () => {
    marked.use(extendedTable());
    expect(marked(trimLines(`
      | H1           | H2      |
      |--------------|---------|
      | This cell    | Cell A  |
      | spans three ^| Cell B  |
      | rows        ^| Cell C  |
    `))).toMatchSnapshot();
  });

  test('Multi-row headers', () => {
    marked.use(extendedTable());
    expect(marked(trimLines(`
      | This header spans two   || Header A |
      | columns *and* two rows ^|| Header B |
      |-------------|------------|----------|
      | Cell A      | Cell B     | Cell C   |
    `))).toMatchSnapshot();
  });

  test('Stops at custom terminators', () => {
    marked.use(extendedTable(['aaaa']));
    expect(marked(trimLines(`
      | Header A | Header B |
      |----------|----------|
      | Cell A   | Cell B   |
      aaaa
    `))).toMatchSnapshot();
  });

  test('Stops at custom multiline terminators', () => {
    marked.use(extendedTable(['aaaa\nbbbb']));
    expect(marked(trimLines(`
      | Header A | Header B |
      |----------|----------|
      | Cell A   | Cell B   |
      aaaa
      bbbb
      cccc
    `))).toMatchSnapshot();
  });

  test('Works with minimal delimiter rows', () => {
    marked.use(extendedTable());
    expect(marked(trimLines(`
      | Header A | Header B | Header C | Header D |
      |-|:-|-:|:-:|
      | Cell A   | Cell B   | Cell C   | Cell D   |
    `))).toMatchSnapshot();
  });

  test('Works with percentage widths', () => {
    marked.use(extendedTable());
    expect(marked(trimLines(`
      | Header A |
      |---10%----|
      | Cell A   |
    `))).toMatchSnapshot();
  });

  test('Works with mix of widths and not', () => {
    marked.use(extendedTable());
    expect(marked(trimLines(`
      | Header A | Header B | Header C |
      |---10%----|----------|--- 50% --|
      | Cell A   | Cell B   | Cell C   |
    `))).toMatchSnapshot();
  });

  test('Works with combined widths and alignment', () => {
    marked.use(extendedTable());
    expect(marked(trimLines(`
      | Header A | Header B | Header C |
      |:---10%---|:-- 20% -:|---50%---:|
      | Cell A   | Cell B   | Cell C   |
    `))).toMatchSnapshot();
  });

  test('Drops empty TRs in multiple tbody Mode', () => {
    marked.use(extendedTable());
    extendedTable().extensions[0].setMode(false);
    expect(marked(trimLines(`
      | Race     | Multi-class options                                         |
      |:---------|:------------------------------------------------------------|
      | Dwarf    | fighter/cleric, fighter/thief                               |
      | Elf      | fighter/mage, fighter/thief, mage/thief,                    |
      |         ^| fighter/mage/thief                                         ^|
      | Gnome    | fighter/thief, fighter/cleric, fighter/illusionist,         | 
      |         ^| thief/cleric, thief/illusionist, cleric/illusionist        ^|
      | Half-elf | fighter/priest, fighter/mage, fighter/thief, ranger/priest, |
      |         ^| mage/priest, mage/thief,                                   ^|
      |         ^| fighter/mage/priest, fighter/mage/thief                    ^|
      | Halfling | fighter/thief                                               |
      | Human    | none                                                        |
    `))).toMatchSnapshot();
  });

  test('Leave empty TRs in single tbody Mode', () => {
    marked.use(extendedTable());
    extendedTable().extensions[0].setMode(true);
    expect(marked(trimLines(`
      | Race     | Multi-class options                                         |
      |:---------|:------------------------------------------------------------|
      | Dwarf    | fighter/cleric, fighter/thief                               |
      | Elf      | fighter/mage, fighter/thief, mage/thief,                    |
      |         ^| fighter/mage/thief                                         ^|
      | Gnome    | fighter/thief, fighter/cleric, fighter/illusionist,         | 
      |         ^| thief/cleric, thief/illusionist, cleric/illusionist        ^|
      | Half-elf | fighter/priest, fighter/mage, fighter/thief, ranger/priest, |
      |         ^| mage/priest, mage/thief,                                   ^|
      |         ^| fighter/mage/priest, fighter/mage/thief                    ^|
      | Halfling | fighter/thief                                               |
      | Human    | none                                                        |
    `))).toMatchSnapshot();
  });
});
