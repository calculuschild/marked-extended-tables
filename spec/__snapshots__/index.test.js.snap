// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extended-table Column Spanning 1`] = `
"<table><thead><tr><th>H1</th>
<th>H2</th>
<th>H3</th>
</tr></thead><tbody><tr><td colspan=3>This cell spans 3 columns</td>
</tr></tbody></table>"
`;

exports[`extended-table Drops empty TRs in multiple tbody Mode 1`] = `
"<table><thead><tr><th align=left>Race</th>
<th align=left>Multi-class options</th>
</tr></thead><tbody><tr><td align=left>Dwarf</td>
<td align=left>fighter/cleric, fighter/thief</td>
</tr></tbody><tbody><tr><td rowspan=2 align=left>Elf </td>
<td rowspan=2 align=left>fighter/mage, fighter/thief, mage/thief, fighter/mage/thief                                         </td>
</tr></tbody><tbody><tr><td rowspan=2 align=left>Gnome </td>
<td rowspan=2 align=left>fighter/thief, fighter/cleric, fighter/illusionist, thief/cleric, thief/illusionist, cleric/illusionist        </td>
</tr></tbody><tbody><tr><td rowspan=3 align=left>Half-elf  </td>
<td rowspan=3 align=left>fighter/priest, fighter/mage, fighter/thief, ranger/priest, mage/priest, mage/thief,                                    fighter/mage/priest, fighter/mage/thief                    </td>
</tr></tbody><tbody><tr><td align=left>Halfling</td>
<td align=left>fighter/thief</td>
</tr></tbody><tbody><tr><td align=left>Human</td>
<td align=left>none</td>
</tr></tbody></table>"
`;

exports[`extended-table Leave empty TRs in single tbody Mode 1`] = `
"<table><thead><tr><th align=left>Race</th>
<th align=left>Multi-class options</th>
</tr></thead><tbody><tr><td align=left>Dwarf</td>
<td align=left>fighter/cleric, fighter/thief</td>
</tr><tr><td rowspan=2 align=left>Elf </td>
<td rowspan=2 align=left>fighter/mage, fighter/thief, mage/thief, fighter/mage/thief                                         </td>
</tr><tr></tr><tr><td rowspan=2 align=left>Gnome </td>
<td rowspan=2 align=left>fighter/thief, fighter/cleric, fighter/illusionist, thief/cleric, thief/illusionist, cleric/illusionist        </td>
</tr><tr></tr><tr><td rowspan=3 align=left>Half-elf  </td>
<td rowspan=3 align=left>fighter/priest, fighter/mage, fighter/thief, ranger/priest, mage/priest, mage/thief,                                    fighter/mage/priest, fighter/mage/thief                    </td>
</tr><tr></tr><tr></tr><tr><td align=left>Halfling</td>
<td align=left>fighter/thief</td>
</tr><tr><td align=left>Human</td>
<td align=left>none</td>
</tr></tbody></table>"
`;

exports[`extended-table Multi-row headers 1`] = `
"<table><thead><tr><th colspan=2 rowspan=2>This header spans two columns <em>and</em> two rows </th>
<th>Header A</th>
</tr><tr><th>Header B</th>
</tr></thead><tbody><tr><td>Cell A</td>
<td>Cell B</td>
<td>Cell C</td>
</tr></tbody></table>"
`;

exports[`extended-table Row Spanning 1`] = `
"<table><thead><tr><th>H1</th>
<th>H2</th>
</tr></thead><tbody><tr><td rowspan=3>This cell spans three  rows        </td>
<td>Cell A</td>
</tr><tr><td>Cell B</td>
</tr><tr><td>Cell C</td>
</tr></tbody></table>"
`;

exports[`extended-table Stops at custom multiline terminators 1`] = `
"<table><thead><tr><th>Header A</th>
<th>Header B</th>
</tr></thead><tbody><tr><td>Cell A</td>
<td>Cell B</td>
</tr></tbody></table><p>aaaa
bbbb
cccc</p>
"
`;

exports[`extended-table Stops at custom terminators 1`] = `
"<table><thead><tr><th>Header A</th>
<th>Header B</th>
</tr></thead><tbody><tr><td>Cell A</td>
<td>Cell B</td>
</tr></tbody></table><p>aaaa</p>
"
`;

exports[`extended-table Works with combined widths and alignment 1`] = `
"<table><thead><tr><th align=left width=10%>Header A</th>
<th align=center width=20%>Header B</th>
<th align=right width=50%>Header C</th>
</tr></thead><tbody><tr><td align=left width=10%>Cell A</td>
<td align=center width=20%>Cell B</td>
<td align=right width=50%>Cell C</td>
</tr></tbody></table>"
`;

exports[`extended-table Works with minimal delimiter rows 1`] = `
"<table><thead><tr><th>Header A</th>
<th align=left>Header B</th>
<th align=right>Header C</th>
<th align=center>Header D</th>
</tr></thead><tbody><tr><td>Cell A</td>
<td align=left>Cell B</td>
<td align=right>Cell C</td>
<td align=center>Cell D</td>
</tr></tbody></table>"
`;

exports[`extended-table Works with mix of widths and not 1`] = `
"<table><thead><tr><th width=10%>Header A</th>
<th>Header B</th>
<th width=50%>Header C</th>
</tr></thead><tbody><tr><td width=10%>Cell A</td>
<td>Cell B</td>
<td width=50%>Cell C</td>
</tr></tbody></table>"
`;

exports[`extended-table Works with percentage widths 1`] = `
"<table><thead><tr><th width=10%>Header A</th>
</tr></thead><tbody><tr><td width=10%>Cell A</td>
</tr></tbody></table>"
`;
